src/log_file.ml

```diff
@@ -0,0 +1,208 @@
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+open! Import
+
+module Make (IO : Io.S) (Key : Data.Key) (Value : Data.Value) = struct
+  module Entry = Data.Entry.Make (Key) (Value)
+
+  module IO = struct
+    include Io.Extend (IO)
+
+    let iter_keys ?min f =
+      let page_size = Int63.(mul Entry.encoded_sizeL (of_int 1_000)) in
+      iter ~page_size ?min (fun ~off ~buf ~buf_off ->
+          let key, _ = Entry.decode_key buf buf_off in
+          f off key;
+          Entry.encoded_size)
+  end
+
+  type t = {
+    io : IO.t;  (** The disk file handler *)
+    append_io : string -> unit;  (** Pre-allocated [IO.append io] closure *)
+    mutable hashtbl : int63 Small_list.t Array.t;
+        (** Hashtable of (key, value) pairs in [io], stored using just their
+            file offsets for memory compactness. Length is always a power of
+            two. *)
+    mutable bucket_count_log2 : int;
+        (** Invariant: equal to [log_2 (Array.length hashtbl)] *)
+    mutable scratch_buf : bytes;
+    mutable cardinal : int;
+  }
+
+  let io t = t.io
+  let cardinal t = t.cardinal
+
+  let clear_memory t =
+    t.hashtbl <- [| Small_list.empty |];
+    t.bucket_count_log2 <- 0;
+    t.cardinal <- 0
+
+  let clear ~generation ?hook ~reopen t =
+    IO.clear ~generation ?hook ~reopen t.io;
+    clear_memory t
+
+  let close t =
+    IO.close t.io;
+    clear_memory t
+
+  let flush ?no_callback ~with_fsync t = IO.flush ?no_callback ~with_fsync t.io
+
+  let key_of_offset t off =
+    let r = IO.read t.io ~off ~len:Key.encoded_size t.scratch_buf in
+    assert (r = Key.encoded_size);
+    fst (Entry.decode_key (Bytes.unsafe_to_string t.scratch_buf) 0)
+
+  let entry_of_offset t off =
+    let r = IO.read t.io ~off ~len:Entry.encoded_size t.scratch_buf in
+    assert (r = Entry.encoded_size);
+    Entry.decode (Bytes.unsafe_to_string t.scratch_buf) 0
+
+  let elt_index t key =
+    (* NOTE: we use the _uppermost_ bits of the key hash to index the bucket
+       array, so that the hashtbl is approximately sorted by key hash (with only
+       the entries within each bucket being relatively out of order). *)
+    let unneeded_bits = Key.hash_size - t.bucket_count_log2 in
+    (Key.hash key lsr unneeded_bits) land ((1 lsl t.bucket_count_log2) - 1)
```

Comment:
If index user's keys aren't uniformly distributed they won't have correct performances with the log. Can you mention this in index's documentation?

Reply:
AFAIK this is actually already true because of the way the fan-out works. There's a warning about it here: https://github.com/mirage/index/blob/201c5a66871c41467af0a360dc0ad5d2c28ac3b1/src/data.ml#L15.